one/src/events.rs

```diff
@@ -48,17 +71,221 @@ async fn slurp(opts: SlurpOpts) -> Result<()> {
     );
 
     let pool = SqlitePool::connect(output_ceramic_path).await.unwrap();
-    let store = SQLiteBlockStore::new(pool).await.unwrap();
+    let block_store = SQLiteBlockStore::new(pool).await.unwrap();
 
     if let Some(input_ceramic_db) = opts.input_ceramic_db {
-        migrate_from_database(input_ceramic_db, store.clone()).await;
+        migrate_from_database(input_ceramic_db, block_store.clone()).await;
     }
     if let Some(input_ipfs_path) = opts.input_ipfs_path {
-        migrate_from_filesystem(input_ipfs_path, store.clone()).await;
+        migrate_from_filesystem(input_ipfs_path, block_store.clone()).await;
+    }
+    Ok(())
+}
+
+async fn validate(opts: ValidateOpts) -> Result<()> {
+    let pool = SqlitePool::from_store_dir(opts.store_dir).await.unwrap();
+    let block_store = SQLiteBlockStore::new(pool.clone()).await.unwrap();
+    let root_store = SQLiteRootStore::new(pool).await.unwrap();
+
+    // Validate that the CID is either DAG-CBOR or DAG-JOSE
+    if (opts.cid.codec() != 0x71) && (opts.cid.codec() != 0x85) {
+        return Err(anyhow!("CID {} is not a valid Ceramic event", opts.cid));
+    }
+
+    // If the CID is a DAG-JOSE, the event is either a signed Data Event or a signed Init Event, both of which can be
+    // validated similarly.
+    if opts.cid.codec() == 0x85 {
+        let controller = validate_data_event_envelope(&opts.cid, &block_store).await;
+        match controller {
+            Ok(controller) => {
+                println!(
+                    "DataEvent({}) validated as authored by Controller({})",
+                    opts.cid, controller
+                )
+            }
+            Err(e) => println!("{} failed with error: {:?}", opts.cid, e),
+        }
+        return Ok(());
+    }
+    // If the CID is a DAG-CBOR, the event is either a Time Event or an unsigned Init Event. In this case, we'll pull
+    // the block from the store and use the presence of the "proof" field to determine whether this is a Time Event or
+    // an unsigned Init Event.
+    //
+    // When validating events from Recon, the block store will be a CARFileBlockStore that wraps the CAR file received
+    // over Recon.
+    if let Some(block) = block_store.get(opts.cid).await? {
```

Comment:
should we error or at least log a warning or something if the block isn't found in the block store?